src/main.ml

```
@@ -83,68 +83,89 @@ let with_attached_vdi sr vdi read_write f =
+                 >>= fun vdi_ref ->
+                 Xen_api.VDI.get_record ~rpc ~session_id ~self:vdi_ref
+                 >>= fun vdi_rec ->
+                 Xen_api.SR.get_uuid ~rpc ~session_id ~self:vdi_rec.API.vDI_SR
+                 >>= fun sr_uuid ->
+                 with_attached_vdi sr_uuid vdi_rec.API.vDI_location (not vdi_rec.API.vDI_read_only)
+                   (fun filename ->
+                      with_block filename (Nbd_lwt_unix.Server.serve t (module Block))
+                   )
+              )
+         )
+         (fun () -> Nbd_lwt_unix.Server.close t)
+    )
+    (ignore_exn channel.close)
 
 let main port =
   let t =
     let sock = Lwt_unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
-    Lwt_unix.setsockopt sock Lwt_unix.SO_REUSEADDR true;
-    let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in
-    Lwt_unix.bind sock sockaddr;
-    Lwt_unix.listen sock 5;
-    let rec loop () =
-      begin
-        Lwt_unix.accept sock
-        >>= fun (fd, _) ->
-        (* Background thread per connection *)
-        let _ =
-          Lwt.catch
-            (fun () -> handle_connection fd)
-            (fun e -> Printf.fprintf stderr "Caught %s\n%!" (Printexc.to_string e); return ())
-          >>= fun () ->
-          Lwt_unix.close fd in
-        return ()
-      end >>= loop
-    in
-    loop ()
+    Lwt.finalize
+      (fun () ->
+         Lwt_unix.setsockopt sock Lwt_unix.SO_REUSEADDR true;
+         let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in
+         Lwt_unix.bind sock sockaddr;
+         Lwt_unix.listen sock 5;
+         let rec loop () =
+           Lwt_unix.accept sock
+           >>= fun (fd, _) ->
+           (* Background thread per connection *)
+           let _ =
+             Lwt.catch
+               (fun () ->
+                  Lwt.finalize
+                    (fun () -> handle_connection fd)
+                    (ignore_exn (fun () -> Lwt_unix.close fd))
```

Same here and below, mention inline why you need to ignore the exception

---

The other reason I ignore exceptions in the `Lwt.finalize` handler is that I found that if any exception happens in it, that exception will overwrite the main exception that triggered it, but I do not want that.

---

Added a comment about the exception raised due to double-closing the fd.

---

```ocaml
utop # Lwt.finalize (fun () -> Lwt.fail_with "e1") (fun () -> Lwt.fail_with "e2");;
Exception: (Failure e2). 
```

---

I've added this reason to the commit message.

---

I knew already, I just think it's better to stick it also in the code in this case

---

Oh, I had overlooked that. Good to have it in the commit message